Feature/sqlalchemy

sampathP opened the pull request:
sqlalchemy implementation for masakari.

Error handling done through sqlalchemy lib and masakari only catch exceptions.
add decorators for session rollback, commit, deadlock handling.

sampathP committed:
pep8 formatting

sampathP committed:
Add init code form sqlalchemy branch form ntt-sic/masakari master

Get the base code for sqlalchemy form ntt-sic/masakari master.
Original work done by toshikazu0314

sampathP committed:
Relocate the controller/db

sampathP committed:
Modify db/api.py to work with thread-safe

Use scoped_session to create thread local sessions.
Modify the get_engine() to create engine form config.

sampathP committed:
Add sqlalchemy test codes
Modify db/api and controller/masakari_controller.py

sampathP committed:
WIP sqlalchemy

sampathP committed:
WIP sqlalchemy

masakari_util update.

sampathP committed:
WIP sqlalchemy

sampathP committed:
WIP sqlalchemy masakari_starter, util

sampathP committed:
WIP sqlalchemy

sampathP committed:
Add db driver conf and remove tests

Add db driver conf and valid drivers are,
'drizzle', 'firebird', 'informix', 'mssql', 'mysql', 'postgresql',
 'sqlite', 'oracle','sybase'

default drive is mysql. sqlite is only for tests.
create_tables.py also supported for sqlite.

sampathP committed:
Refine code

sampathP committed:
Add sqlalchemy Decorators

sampathP committed:
Add decorators and refine the code

sampathP committed:
Resolve merge conflicts

sampathP committed:
Resolve merge conflicts

sampathP commented:
Conflicts are resolved.
Merge button does not show up... :imp: 
But still can merge via CLI .. :bulb:

sampathP committed:
Fix: debian pack miss db directory

muroi commented on masakari-controller/controller/masakari_util.py:
```diff
@@ -361,54 +348,42 @@ def update_notification_list_db(self, key, value, notification_id):
-        ## TODO:
-        ## Get the cursor from Func Args and remove this MySQLdb.connect
+        # TODO:
+        # Get the cursor from Func Args and remove this MySQLdb.connect
```
The TODO seems to be unnecessary.

muroi commented on masakari-controller/controller/masakari_worker.py:
```diff
@@ -570,7 +540,7 @@ def msg_filter(return_message, ignore_message_list):
         return status
 
     def _do_action_db(self, config, sql):
```
When moving MySQLdb to sqlalchemy, is it the method needed?

muroi commented on masakari-controller/db/api.py:
```diff
@@ -0,0 +1,524 @@
+
+@_retry_on_deadlock
+@_session_handle
+def delet_expired_notification(session, progress, update_at, delete_at, id):
```
IMO, the method don't need to have "progress" arg since following query don't use it.

muroi commented:
The PR looks good to me. If possible, could you check my inline comment?

sampathP committed:
Fix for @muroi comments

sampathP commented:
@muroi Thank you for review.
Please find the fix in previous patch

muroi commented:
Thank you for quick updating! I want to merge it, but I hit some following issues when I deploy this with masakari-deploy. Do you have any work for these? 
1. masakari-controller/db directory is not put under /opt/masakari/  by deb installer
2. I was required to additionally install sqlalchemy_utils ($ sudo pip install sqlalchemy_utils) even after I installed openstack

Additionally, I hit unexpected error during testing. Sorry, I should've tested it and comment it first.
1.  in recovering instance failure, I received the log in masakari-controller.log and instance is not rescued.
   <_MainThread(MainThread, started 140390764574528)> must be string, not None at  File "/opt/masakari/controller/masakari_util.py", line 205, in insert_notification_list_db
2. in recovering host failure, I received the log in masakari-controller.log and instance is not rescued.
   <Thread(Thread-4, started 139735165568768)> 'unicode' object has no attribute 'rollback' at File "/opt/masakari/controller/masakari_starter.py", line 163, in _create_vm_list_db_for_failed_host( notification_uuid)

sampathP committed:
Add db to /opt/masakari in deb packaging

sampathP committed:
Fix missing parameter "session" for
dbapi.get_one_vm_list_by_uuid_and_progress_create_at_last
at masakari_starter.py

Add try-catch to datetime.strptime in,
masakari_util.py
to avoid the error when strping the
"None" time stamps.

sampathP commented:
Sorry for the inconvenience.
- Add db to /opt/masakari
- I will push a patch to masakari-deploy to install sqlalchemy_utils on masakari_controller.
  Errors,
- Cause was, datetime.strptime() cant convert None to given format. 
  Fix it by checking  the exceptions of type-error, or value-error and return None if any.
  Otherwise return the datetime.strptime().
- Cause was parameter miss. parameter "session" was not given to db.api and  _session_handle() try to rollback() with the only parameter given which was  "notification_uuid".

sampathP committed:
Fix the wrong dpkg-statoverride description for postrm in deb packaging

sampathP committed:
Fix errors on sqlalchemy related codes

muroi commented:
Thank you @sampathP . I hit same bug you resolved by last commit.  Eventually, the PR works well. So I merged it manually since there is a merge conflict.

Close #5
